File: Python/day_2.py
# // Differences = 0 => val[1] + 3 (3 combos)
# //               1 => val[1] + 6 (2 combos)
# //               2 => val[1]     (1 combo)
# //              -1 => val[1]     (2 combos)
# //              -2 => val[1] + 6 (1 combo)

def day_2():

    total_a = 0
    total_b = 0
    pattern = [1, 2, 3]
    pairs = {0: 3, 1: 6, 2: 0, -1: 0, -2: 6}

    with open("../inputs/day_2.txt", "r") as f:
        lines = f.readlines()

    # with open("testing.txt", "r") as f:
    #     lines = f.readlines()

    for line in lines:
        a, b = ord(line[0]) - 64, ord(line[2]) - 87

        total_a += (b + pairs[b - a])
        total_b += ((pattern[(a - 2) % 3]) if b == 1
                    else (pattern[a % 3] + 6) if b == 3
                    else (a + 3))

    print(f"Total A: {total_a}\nTotal B: {total_b}")


if __name__ == "__main__":
    day_2()

File: Rust/src/day_2.rs
pub fn solution() {

    let input = include_str!("./input.txt")
        .lines()
        .map(|l| l.split_once(" ").unwrap())
        .map(|(a, b)|
            (a.to_string().as_bytes().to_owned()[0] - 64, b.to_string().as_bytes().to_owned()[0] - 87))
        .collect::<Vec<(u8, u8)>>();

    // println!("{:?}", input);
    // Rock: 1
    // Paper: 2
    // Scissors: 3
    // Differences = 0 => val[1] + 3 (3 combos)
    //               1 => val[1] + 6 (2 combos)
    //               2 => val[1]     (1 combo)
    //              -1 => val[1]     (2 combos)
    //              -2 => val[1] + 6 (1 combo)

    // part 1
    let mut total_1: i32 = input.iter().fold(0, |acc, (a, b)|
    match *b as i32 - *a as i32 {
         0 => acc + 3 + *b as i32,
         1 => acc + 6 + *b as i32,
         2 => acc + *b as i32,
        -1 => acc + *b as i32,
        -2 => acc + *b as i32 + 6,
        _ => acc
    });

    // 1 -> lose    *a == 1: 3, *a == 2: 1, *a == 3: 2
    // 2 -> draw    3 + *a
    // 3 -> win     *a == 1: 2, *a == 2: 3, *a == 3: 1 (all plus 6)

    let mut total_2: i32 = input.iter().fold(0, |acc, (a, b)|
        match *b as i32 {
            1 => acc + match *a as i32 {
                1 => 3,
                2 => 1,
                3 => 2,
                _ => 0
            },
            2 => acc + *a as i32 + 3,
            3 => acc + 6 + match *a as i32 {
                1 => 2,
                2 => 3,
                3 => 1,
                _ => 0
            },
            _ => acc
        });

    println!("The total for part 1: {} \n", total_1);
    println!("The total for part 2: {} \n", total_2);
}